src/jdk.zipfs/share/classes/jdk/nio/zipfs/ZipFileSystem.java

```diff
@@ -3085,14 +3085,26 @@ private void readExtra(ZipFileSystem zipfs) throws IOException {
                 int sz = SH(extra, pos + 2);
                 pos += 4;
                 if (pos + sz > elen) {        // invalid data
-                    throw new ZipException("Invalid CEN header (invalid zip64 extra data field size)");
+                    throw new ZipException(String.format(
+                            "Invalid CEN header (invalid extra data field size for " +
+                                    "tag: 0x%04x size: %d)",
+                            tag, sz));
                 }
                 switch (tag) {
                 case EXTID_ZIP64 :
                     // Check to see if we have a valid block size
                     if (!isZip64ExtBlockSizeValid(sz)) {
                         throw new ZipException("Invalid CEN header (invalid zip64 extra data field size)");
                     }
+                    // if ZIP64_EXTID blocksize == 0, validate csize, size and
+                    // locoff to make sure they do not == ZIP64_MINVAL
+                    if (sz == 0) {
+                        if ( csize == ZIP64_MINVAL || size == ZIP64_MINVAL ||
+                                locoff == ZIP64_MINVAL) {
+                            throw new ZipException("Invalid CEN header (invalid zip64 extra data field size)");
+                        }
+                        break;
+                    }
                     if (size == ZIP64_MINVAL) {
```

Comment:
Note that we always increase "pos" only in case of "_MINVAL". If the values of size and csize are correct/valid in the "body" of the zip file and only locoff is negative then we should skip two fields in the extra block and read the third one. Otherwise, we may read some random values and throw an exception.

Reply:
I am not sure I (quite) understand your question completely..

How ZIpFS::readExtra  has navigated these fields has not changed

 If you have a tool that creates a zip/jar that demonstrates an issue that might need further examination, please provide a test case, the tool that created the zip/jar in question and open a new bug.

Reply:
The 8302483 changed this code to throw an exception, this is why I am looking into it.
You can compare the code in this file and the same code in the ZipFile in the checkZip64ExtraFieldValues method or the code in the ZipEntry#setExtra0, where we do not increase the "off" but instead checks for "off+8" or "off + 16". So if we need to read only the third field we should read "pos+16" but for the current implementation we will read it at "pos+0" since the pos was not bumped by the code for two other fields.